Add anonymous_user_id setting description.

kmadejski opened the pull request:
This PR contains short description about `anonymous_user_id` setting. 

These changes cover approach valid for eZ Platform 1.7 and lower. 
It requires additional changes for 1.13 and up.

kmadejski commented:
@andrerom could you take a look, please?

andrerom commented:
Looks ok for anonymous_user_id, but didn't we have a need to also document how this can be used with Varnish in regards to fos hash lookups?

kmadejski commented:
Thanks. For now, I've added the only small tip on Varnish page, but of course, I can improve it if you think that possibility should be wider documented.

andrerom commented:
Well, wouldn't user need to adapt VCL to be able to use it? Doc changes here seems to imply it will work out of the box, was that the conclusion from recent issue on this?

kmadejski commented:
@andrerom yes, you are right. It should be covered also. I'm going to add this use-case.

kmadejski commented:
@andrerom I've added a section about using anonymous_user_id with Varnish, could you review once again, please?

DominikaK commented on docs/guide/best_practices.md:
```diff
@@ -151,3 +151,18 @@ ezpublish:
 **Order of precedence is not modified.**
+
+### `anonymous_user_id`
+When a user that is not logged in is accessing an eZ Platform installation, the request will be handled with the permissions of the anonymous user. You can select any user on your system to represent the anonymous user.
```
`When a user who is...` and `select any user in your system`

DominikaK commented on docs/guide/best_practices.md:
```diff
@@ -151,3 +151,18 @@ ezpublish:
+
+### `anonymous_user_id`
+When a user that is not logged in is accessing an eZ Platform installation, the request will be handled with the permissions of the anonymous user. You can select any user on your system to represent the anonymous user.
+For instance, you can create also separated anonymous user group per siteaccess. Then, you can create the new user, assign him to the new group and make all permissions depend on siteaccess.
```
`For instance, you can also create a separate anonymous user group per siteaccess. Then, you can create the new user, assign them to the new group and make all permissions depend on siteaccess.`

DominikaK commented on docs/guide/best_practices.md:
```diff
@@ -151,3 +151,18 @@ ezpublish:
+For instance, you can create also separated anonymous user group per siteaccess. Then, you can create the new user, assign him to the new group and make all permissions depend on siteaccess.
+Take care to give the correct permissions to the anonymous user since these permissions are available for all users on your site. 
+
+You can set `anonymous_user_id` in the way like:
```
`in the way like` -> `the following way` or `like this`

DominikaK commented on docs/guide/http_cache.md:
````diff
@@ -730,11 +737,167 @@ set req.http.X-User-Hash = "b1731d46b0e7a375a5b024e950fdb8d49dd25af85a5c7dd5116a
+        
+```
+
+You need to get the new X-User-Hash for every new anonymous user / anonymous users group. It is done in exactly same way as described above ([User hash generation](http_cache.md#user-hash-generation)), but you have to remember to use siteaccess URI element and additional header which tells eZ which URI should take into account. For instance for `eng` siteaccess:
````
`It is done in exactly the same way`
`remember to use the siteaccess URI element and an additional header`
`which URI it should take into account`

DominikaK commented on docs/guide/http_cache.md:
```diff
@@ -730,11 +737,167 @@ set req.http.X-User-Hash = "b1731d46b0e7a375a5b024e950fdb8d49dd25af85a5c7dd5116a
+  
+  `curl -I -H "Accept: application/vnd.fos.user-context-hash" -H "x-fos-original-url: /nor/" http://<your-domain.com>/nor/_fos_user_context_hash`
+
+Lets assume, that the new X-User-Hashes are:
```
`Let's assume that`

andrerom commented on docs/guide/http_cache.md:
```diff
@@ -730,11 +737,167 @@ set req.http.X-User-Hash = "b1731d46b0e7a375a5b024e950fdb8d49dd25af85a5c7dd5116a
+            }
+            // Pre-authenticate request to get shared cache, even when authenticated
+            else {
+                set req.http.x-fos-original-url    = req.url;
```
Afaik we don't need to duplicate this section, the VCL session cookie pattern should be able to handle `_whatever` on session name.

Also lets move this syntax to be for ezplatform-http-cache VCL syntax, so it's the same across 1.7(.7) / 1.13 / 2.0 versions. In this xkey VCL flavor session cookie logic is already handled in `vcl_recv()` so less complicated logic in `ez_user_context_hash ()` to explain how to modify.

Optional: Move the VCL part of the doc to 1.13 to avoid having to explain which VCL type example is for, most users are moving to 1.13 / 2.0 anyway.

kmadejski commented on docs/guide/http_cache.md:
````diff
@@ -337,6 +337,13 @@ ezpublish:
+framework:
+    trusted_proxies: [127.0.0.1,193.22.44.22]
+```
+
````
Yes, you are right. It is not necessary even in 2.8.x if _SYMFONY_TRUSTED_PROXIES_ is set properly. I'm going to remove this part.

kmadejski commented on docs/guide/http_cache.md:
```diff
@@ -303,7 +303,7 @@ Somehow we need to tell php process that we are behind a Varnish proxy and not t
     # Configure IP of your Varnish server to be trusted proxy
     # Replace fake IP address below by your Varnish IP address
-    SetEnv SYMFONY_TRUSTED_PROXIES "193.22.44.22"
+    SetEnv SYMFONY_TRUSTED_PROXIES "127.0.0.1,193.22.44.22"
```
Yesterday we got ticket related to this setting. Mainly, it is required when HTTPS is in use and `{{render_esi(controller(...))}}` or `hinclude.js` is involved.

andrerom replied:
Thats not an explanation ;) In order to explain why this is needed we also need to understand why it is like that. Is the example regarding cases where one of the proxies run on same machine, or is it caused by web server acting on the forwarding headers also so we need to tell symfony to trust it?

kmadejski replied:
I've spent some time trying to find out the use cases when it is necessary, and honestly, I'm lost a bit. I can't find the correlation between this setting and some of the issues which our customers experienced in the past, even if setting this value solved problems. I will dig into this matter and will be back to this part of documentation as soon as possible with full and clear understanding.

andrerom replied:
@kmadejski I guess you can drop this change now given Symfony will release your fix in next release ;)

kmadejski commented:
@DominikaK @MagdalenaZuba just for your information, I'm still waiting for the customer's feedback about this solution. We decided to include this in our documentation because of his use-case, so basically I need final confirmation that this approach works as expected. I'll finish this PR as soon as possible.

kmadejski commented:
@andrerom I've simplified the example code and clean-up the whole PR according to the `TRUSTED_PROXIES` change in Symfony.
@DominikaK I've rebased this PR, so there are no conflicts. Please review again.

andrerom commented on docs/guide/http_cache.md:
```diff
@@ -501,6 +501,118 @@ set req.http.X-User-Hash = "b1731d46b0e7a375a5b024e950fdb8d49dd25af85a5c7dd5116a
 
+##### New anonymous X-User-Hash based on `anonymous_user_id` setting
+
+Since you can configure anonymous user per SiteAccess, you need to change the default `.vcl` template to be able to benefit from this functionality. This example assumes that your SiteAccesses are matched using `URLElement` matcher. It works also with other SiteAccess matchers. As Varnish is not SiteAccess-aware itself, it is recommended to set also different session cookie names per SiteAccess.
```
Maybe make it clear it's in cases where site access is matched by url we need to do this. In case of matching by host or port I think Varnish will keep the cache items separated.

andrerom replied:
Also, why do we suggest using different session cookie name btw?

kmadejski replied:
Updated the part about matchers.

> Also, why do we suggest using different session cookie name btw?
I kept it from the customer's solution, but TBH I can see now, that this is not necessary. Therefore, I've removed it also.
